grr/server/grr_response_server/flows/general/osquery.py

```diff
@@ -67,13 +100,94 @@ def _UpdateProgress(
     self.state.progress.partial_table = _GetTruncatedTable(responses)
     self.state.progress.total_row_count = _GetTotalRowCount(responses)
 
+  def _GetPathSpecsToCollect(
+      self,
+      responses: flow_responses.Responses[rdf_osquery.OsqueryResult],
+  ) -> List[rdf_paths.PathSpec]:
+    if len(self.args.file_collect_columns) == 0:
+      return []
+
+    total_row_count = _GetTotalRowCount(responses)
+    if total_row_count > FILE_COLLECT_LIMIT_MAX_ROWS:
+      message = (f"Requested file collection on a table with {total_row_count} "
+          f"rows, but the limit is {FILE_COLLECT_LIMIT_MAX_ROWS} rows")
+      raise FileCollectionLimitsExceeded(message)
+
+    file_names = []
+    for osquery_result in responses:
+      for column in self.args.file_collect_columns:
+        file_names.extend(osquery_result.table.Column(column))
+
+    return [rdf_paths.PathSpec.OS(path=file_name) for file_name in file_names]
+
+  def _FileCollectionFromColumns(
+      self,
+      responses: flow_responses.Responses[rdf_osquery.OsqueryResult],
+  ) -> None:
+    pathspecs = self._GetPathSpecsToCollect(responses)
+
+    stub = server_stubs.GetFileStat
+    for pathspec in pathspecs:
+      request = rdf_client_action.GetFileStatRequest(pathspec=pathspec)
+      self.CallClient(
+          stub,
+          request,
+          next_state=compatibility.GetName(self._StatForFileArrived))
+
+  def _StatForFileArrived(
+      self,
+      responses: flow_responses.Responses[rdf_client_fs.StatEntry],
+  ) -> None:
+    if not responses.success:
+      # TODO(simstoykov): Take note of this and report to the user with other
+      # flow results, instead of failing the flow.
+      raise flow_base.FlowError(
+        f"Error when attempted to stat file: {responses.status}")
+
+    if len(responses) != 1:
+      raise NotImplementedError(
```

Comment:
`ValueError` feels more appropriate, `NotImplementedError` is more for TODO-like things.